modules/certificate/certificate.service.js

```diff
@@ -42,38 +49,69 @@ class CertificatesService {
     return certificate;
   }
 
-  async addCertificate(name, value, id) {
-    const certificateExists = await Certificate.findOne({ name });
+  async generateCertificate(certificateData, userId, userRole) {
+    const randomNum = randomInt(100000, 999999);
+    const name = `hor${randomNum}`;
 
-    if (certificateExists) {
-      throw new RuleError(CERTIFICATE_ALREADY_EXISTS, BAD_REQUEST);
+    certificateData.name = name;
+
+    if (userRole === USER) {
+      certificateData.ownedBy = userId;
+    } else if (userRole === ADMIN || userRole === SUPERADMIN) {
+      certificateData.createdBy = userId;
     }
 
-    return new Certificate({ name, value, createdBy: id }).save();
+    return new Certificate(certificateData).save();
   }
 
-  async updateCertificate(name) {
-    const updatedCertificate = await Certificate.findOneAndUpdate(
+  async addCertificate(name, userId, userEmail) {
+    const certificateExists = await Certificate.findOne({ name });
+
+    if (
+      !certificateExists ||
+      certificateExists.isUsed ||
+      certificateExists.isExpired
+    ) {
+      throw new RuleError(CERTIFICATE_NOT_VALID, BAD_REQUEST);
+    }
+
+    if (certificateExists.ownedBy) {
+      throw new RuleError(CERTIFICATE_HAVE_OWNER, BAD_REQUEST);
+    }
+
+    return Certificate.findOneAndUpdate(
       { name },
-      { isUsed: true },
+      { email: userEmail, ownedBy: userId },
       { new: true }
     ).exec();
+  }
+
+  async updateCertificate(name) {
+    const certificateExists = await Certificate.findOne({ name }).exec();
```

Comment:
you use Certificate.findOne twice. put in separate method please , call getCertificateByParams and use with this.getCertificateByParams

Reply:
done, thanks for an idea